metacat-connector-polaris/src/main/java/com/netflix/metacat/connector/polaris/store/repos/PolarisTableCustomRepositoryImpl.java

```diff
@@ -0,0 +1,78 @@
+package com.netflix.metacat.connector.polaris.store.repos;
+
+import javax.persistence.PersistenceContext;
+import javax.persistence.EntityManager;
+import javax.persistence.Query;
+
+import com.netflix.metacat.connector.polaris.store.entities.PolarisTableEntity;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.data.domain.Slice;
+import org.springframework.data.domain.SliceImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation for Custom JPA repository implementation for storing PolarisTableEntity.
+ */
+@Repository
+public class PolarisTableCustomRepositoryImpl implements PolarisTableCustomRepository {
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    private Slice<PolarisTableEntity> findAllTablesByDbNameAndTablePrefixForCurrentPage(
+        final String dbName, final String tableNamePrefix, final Pageable page) {
+
+        // Generate ORDER BY clause
+        String orderBy = "";
+        if (page.getSort().isSorted()) {
+            orderBy = page.getSort().stream()
+                .map(order -> order.getProperty() + " " + order.getDirection())
+                .collect(Collectors.joining(", "));
+            orderBy = " ORDER BY " + orderBy;
+        }
+
+        final String sql = "SELECT t.* FROM TBLS t "
+            + "WHERE t.db_name = :dbName AND t.tbl_name LIKE :tableNamePrefix" + orderBy;
+        final Query query = entityManager.createNativeQuery(sql, PolarisTableEntity.class);
+        query.setParameter("dbName", dbName);
+        query.setParameter("tableNamePrefix", tableNamePrefix + "%");
+        query.setFirstResult(page.getPageNumber() * page.getPageSize());
+        query.setMaxResults(page.getPageSize() + 1); // Fetch one extra result to determine if there is a next page
+        final List<PolarisTableEntity> resultList = query.getResultList();
+        // Check if there is a next page
+        final boolean hasNext = resultList.size() > page.getPageSize();
+        // If there is a next page, remove the last item from the list
+        if (hasNext) {
+            resultList.remove(resultList.size() - 1);
+        }
+        return new SliceImpl<>(resultList, page, hasNext);
+    }
+
+    @Override
+    @Transactional
+    public List<PolarisTableEntity> findAllTablesByDbNameAndTablePrefix(
+        final String dbName, final String tableNamePrefix, final int pageFetchSize) {
+        Pageable page = PageRequest.of(0, pageFetchSize, Sort.by("tbl_name").ascending());
+        entityManager.createNativeQuery("SET TRANSACTION AS OF SYSTEM TIME follower_read_timestamp()")
```

Comment:
The `entityManager` is an instance variable in this class so if multiple threads are all running for this api at the same time, would there be situations where they overwrite the timestamp set by each other, since `follower_read_timestamp()` is a crdb function that is evaluated to a value upon execution

Reply:
Since the entityManager is annotated with `@PersistenceContext` it's scoped within a transaction.

https://www.baeldung.com/jpa-hibernate-persistence-context

Reply:
Thanks for the info and reference, upon closer look seems like spring will inject a special proxy for the `entityManager` in this setup so indeed should be safe
https://github.com/spring-projects/spring-framework/blob/main/spring-orm/src/main/java/org/springframework/orm/jpa/SharedEntityManagerCreator.java